README.md
# Overview

The Carbon.Windows.Installer module is a Windows-only module that has functions for reading and installing Windows MSI
files/packages, and for replicating Windows' "Programs and Features"/"Apps and Features" graphical user interface but
with objects and text.

# System Requirements

* Windows PowerShell 5.1 on .NET 4.5.2+
* PowerShell 6.2, 7.1, or 7.2
* Windows Server 2012 R2+ or Windows 8.1+

# Installing

To install globally:

```powershell
Install-Module -Name 'Carbon.Windows.Installer'
Import-Module -Name 'Carbon.Windows.Installer'
```

To install privately:

```powershell
Save-Module -Name 'Carbon.Windows.Installer' -Path '.'
Import-Module -Name '.\Carbon.Windows.Installer'
```

# Commands

By default, commands have a `C` prefix. This can be changed with `Import-Module` cmdlet's `Prefix` parameter.

* `Get-CInstalledProgram`: reads the registry to determine all the programs installed on the local computer. Returns
an object for each program. Is an object-based and text-based version of Windows' "Programs and Features"/"Apps and
Features" GUI.
* `Get-CMsi`: reads an MSI file and returns an object exposing the MSI's internal tables, like product name, product
code, product version, etc. Let's you inspect an MSI for its metadata without installing it. Can also download the MSI
file first. This function requires Windows PowerShell or PowerShell 7.1+ on Windows.
* `Install-CMsi`: installs a program from an MSI file, or other file that can be installed by Windows. Can also 
download the MSI file to install. This function requires Windows PowerShell or PowerShell 7.1+ on Windows.


CHANGELOG.md

# 2.0.0

## Upgrade Instructions

The `DefaultCommandPrefix` property doesn't work as we expected, so we've removed it and added an explicit `C` prefix
on each command. Add a `C` prefix to each function name on usages of `Import-Module` cmdlet's `Function` parameter.

## Changed

The module now exposes its commands with an explict `C` prefix instead of using the `DefaultCommandPrefix` parameter.

# 1.0.0

The `Carbon.Windows.Installer` module was created from functions in the `Carbon` module. These release notes assume
you're migrating from [Carbon](http://get-carbon.org). If you're not migrating, you can ignore these release notes.

## Migration Instructions

If migrating from Carbon, you'll need to make the following changes:

### General

* `Carbon.Windows.Installer` requires Windows PowerShell 5.1 or PowerShell 6.2+ on Windows.

### Get-CMsi

* Rename usages of `Get-Msi` to `Get-CMsi` (the `C` prefix is now required).
* Rename `Properties` property usages to `Property` on objects returned from `Get-CMsi`. The `Properties` property on
  objects returned from `Get-CMsi` was renamed to `Property`.
* Change property lookups on the objects returned from `Get-CMsi` from `$msiInfo.Properties[KEY]` to
  `$msiInfo.GetPropertyValue(KEY)`. The `Properties` property is now named `Property` and is an array of
  `Carbon.Windows.Installer.Records.Property` objects (instead of a hashtable) and we added a `GetPropertyValue` method
  to lookup specific property values.
* Remove usages of the `[Carbon.Msi.MsiInfo]` type. It was removed. `Get-CMsi` now returns `[PSObject]` objects with a
  pstypename of `[Carbon.Windows.Installer.MsiInfo]`.

### Get-CProgramInstallInfo

* Rename usages of `Get-CProgramInstallInfo` to `Get-CInstalledProgram`. We renamed `Get-CProgramInstallInfo` to
  `Get-CInstalledProgram`. The `C` prefix is now required.
* Remove usages of the `[Carbon.Computer.ProgramInstallInfo]` type. `Get-CInstalledProgram` now returns `[PSObject]`
objects with a pstypename of `[Carbon.Windows.Installer.ProgramInfo]`.
* Add `-ErrorAction Ignore` or `-ErrorAction SilentlyContinue` to usages of `Get-CInstalledProgram` that handle if no
object is returned. `Get-CInstalledProgram` now writes an error if a program isn't installed.

### Install-CMsi

* Remove usages of the `-Quiet` switch. The `-Quiet` parameter was removed.

## Release Notes

### Added

* You can now change the default `C` prefix when importing `Carbon.Windows.Installer` with the `Import-Module` cmdlet's
`-Prefix` parameter.
* `Get-CMsi`: returned objects now have a `GetPropertyValue([String])` method for getting property values.
* `Get-CInstalledProgram` (i.e. `Get-CProgramInstallInfo`) now writes an error when a program isn't found.
* `Get-CMsi` can now read records from *all* of an MSI's internal database tables. By default, only records from the
`Product` and `Feature` tables are returned. The objects returned by `Get-CMsi` now have a `TableNames` property, which
is the list of all the tables in the MSI's database, and `Tables`, which contains properties for each included table.
To include a table, pass its name (or a wildcard) to the new `IncludeTable` parameter.
* `Get-CMsi` can now download MSI files. Pass the URL to the MSI file to the new `Url` parameter. Use the `Path` of the
return object to get the file's location. Use the new `OutputPath` parameter to save the installer to a specific
directory or to a specific file.
* When an installer fails, `Install-CMsi` leaves a debug log of the installation in the user's temp directory. The
file's name begins with the installer's file name followed by a random file name and has a `.log` extension.
* `Install-CMsi` can now download an MSI file to install. Pass the URL to the installer to the `Url` parameter, the
file's SHA256 checksum to the `Checksum` parameter, its product name to the `ProductName` parameter, and its
product code to the `ProductCode` parameter. The MSI file will only be downloaded if it isn't installed. You can get the MSI's checksum with PowerShell's `Get-FileHash` cmdlet. You can get the MSI's product name and code with this module's `Get-CMsi` function.
* Added `ArgumentList` parameter to `Install-CMsi` to pass additional arguments to `msiexec.exe` when installing an MSI.
* Added `DisplayOption`, `LogOption`, and `LogPath` parameters to allow users to control the display and logging options passed to `msiexec.exe`.

### Changed

* Minimum system requirements are now Windows PowerShell 5.1 on .NET 4.5.2+ or PowerShell 6.2+ on Windows. The
`Get-CMsi` and `Install-CMsi` functions don't work in PowerShell 6.2.
* `Get-CMsi`: The `Properties` property on returned objects renamed to `Property`.
* `Get-CMsi`: The `Properties` property on returned objects is no longer a hashtable but an array of
`[Carbon.Windows.Installer.Records.Property]` objects. To find the value of a property, use the
`GetPropertyValue([String])` method.
* `Get-CMsi` now returns `[PSObject]` objects (instead of `[Carbon.Msi.MsiInfo]` objects) with a pstypename of
`[Carbon.Windows.Installer.MsiInfo]`.
* `Get-CProgramInstallInfo` renamed to `Get-CInstalledProgram`.
* `Get-CInstalledProgram` no longer writes errors when not running as an administrator.
* `Get-CInstalledProgram` now returns `[PSObject]` objects (instead of `[Carbon.Computer.ProgramInstallInfo]` objects)
with a pstypename of `[Carbon.Windows.Installer.ProgramInfo]`.

## Removed

* `Get-Msi` backward-compatible shim function. Use `Get-CMsi` instead.
* `Get-ProgramInstallInfo` backward-compatible shim function. Use `Get-CInstalledProgram` instead.
* `[Carbon.Msi.MsiInfo]` .NET object.
* `[Carbon.Computer.ProgramInstallInfo]` .NET object.
* `Install-Msi` backwards-compatible shim function. Use `Install-CMsi` instead.


Carbon.Windows.Installer/Functions/Install-Msi.ps1

function Install-Msi
{
    <#
    .SYNOPSIS
    Installs an MSI.

    .DESCRIPTION
    `Install-CMsi` installs software from an MSI file, without displaying any user interface. Pass the path to the MSI
    to install to the `Path` property. The `Install-CMsi` function reads the product name code from the MSI file, and
    does nothing if a program with that product code is already installed. Otherwise, the function runs the installer in
    quiet mode (i.e. no UI is visible) with `msiexec`. All the program's features will be installed with their default
    values. You can control the installer's display mode with the `DisplayMode` parameter: set it to `Passive` to show a
    UI with just a progress bar, or `Full` to show the UI as-if the user double-clicked the MSI file.

    `Install-CMsi` can also download an MSI and install it. Pass the URL to the MSI file to the `Url` parameter. Pass
    the MSI file's SHA256 checksum to the `Checksum` parameter. (Use PowerShell's `Get-FileHash` cmdlet to get the
    checksum.) In order avoid downloading an MSI that is already installed, you must also pass the MSI's product name to
    the `ProductName` parameter and its product code to the `ProductCode` parameter. Use this module's `Get-CMsi`
    function to get an MSI file's product metadata.

     If the install fails, the function writes an error and leaves a debug-level log file in the current user's temp
     directory. The log file name begins with the name of the MSI file name, then has a `.`, then a random file name
     (e.g. `xxxxxxxx.xxx`), then ends with a `.log` extension. You can customize the location of the log file with the
     `LogPath` parameter. You can customize logging options with the `LogOption` parameter. Default log options are
    `!*vx` (log all messages, all verbose message, all debug messages, and flush each line to the log file as it is
    written).

    If you want to install the MSI even if it is already installed, use the `-Force` switch. For downloaded MSI files,
    this will cause the file to be downloaded every time `Install-CMsi` is run.

    You can pass additional arguments to `msiexec.exe` when installing the MSI file with the `ArgumentList` parameter.

    Requires Windows PowerShell 5.1 or PowerShell 7.1+ on Windows.

    .EXAMPLE
    Install-CMsi -Path '.\Path\to\installer.msi'
    
    Demonstrates how to install a program with its MSI.

    .EXAMPLE
    Get-ChildItem *.msi | Install-CMsi

    Demonstrates that you can pipe file objects to `Install-CMsi`.

    .EXAMPLE
    Install-CMsi -Path 'installer.msi' -Force

    Demonstrates how to re-install an MSI file even if it's already installed.

    .EXAMPLE
    Install-CMsi -Url 'https://example.com/installer.msi' -Checksum '63c34def9153659a825757ec475a629dff5be93d0f019f1385d07a22a1df7cde' -ProductName 'Carbon Test Installer' -ProductCode 'e1724abc-a8d6-4d88-bbed-2e077c9ae6d2'

    Demonstrates that `Install-CMsi` can download and install MSI files.
    #>
    [CmdletBinding(SupportsShouldProcess, DefaultParameterSetName='ByPath')]
    param(
        # The path to the installer to run. Wildcards supported.
        [Parameter(Mandatory, ValueFromPipeline, ValueFromPipelineByPropertyName, ParameterSetName='ByPath')]
        [Alias('FullName')]
        [string[]] $Path,

        # The URL to an installer to download and install. Requires the `Checksum` parameter to ensure the correct file
        # was downloaded.
        [Parameter(Mandatory, ParameterSetName='ByUrl')]
        [Uri] $Url,

        # Used with the `Url` parameter. The SHA256 hash the downloaded installer should have. Case-insensitive.
        [Parameter(Mandatory, ParameterSetName='ByUrl')]
        [String] $Checksum,

        # The product name of the downloaded MSI. Used to determine if the program is installed or not. Used with the
        # `Url` parameter. The installer is only downloaded if the product is not installed or the `-Force` switch is
        # used. Use the `Get-CMsi` function to get the product code of an MSI.
        [Parameter(Mandatory, ParameterSetName='ByUrl')]
        [String] $ProductName,
        
        # The product code of the downloaded MSI. Used to determine if the program is installed or not. Used with the
        # `Url` parameter. The installer is only downloaded if the product is not installed or the `-Force` switch is
        # used. Use the `Get-CMsi` function to get the product code from an MSI.
        [Parameter(Mandatory, ParameterSetName='ByUrl')]
        [Guid] $ProductCode,
        
        # Install the MSI even if it has already been installed. Will cause a repair/reinstall to run.
        [Switch] $Force,

        # Controls how the MSI UI is displayed to the user. The default is `Quiet`, meaning no UI is shown. Valid values
        # are `Passive`, a UI showing a progress bar is shown, or `Full`, the UI is displayed to the user as if they
        # double-clicked the MSI file.
        [ValidateSet('Quiet', 'Passive', 'Full')]
        [String] $DisplayMode = 'Quiet',

        # The logging options to use. The default is to log all information (`*`), log verbose output (`v`), log extra
        # debugging information (`x`), and to flush each line to the log (`!`).
        [String] $LogOption = '!*vx',

        # The path to the log file. The default is to log to a file in the temporary directory and delete the log file
        # unless the installation fails. The default log file name begins with the name of the MSI file name, then
        # has a `.`, then a random file name (e.g. `xxxxxxxx.xxx`), then ends with a `.log` extension.
        [String] $LogPath,

        # Extra arguments to pass to the installer. These are passed directly after the install option and path to the
        # MSI file. Do not pass any install option, display option, or logging option in this parameter. Instead, use
        # the `DisplayMode` parameter to control display options, the `LogOption` parameter to control logging options,
        # and the `LogPath` parameter to control where the installation log file is saved.
        [String[]] $ArgumentList
    )

    process
    {
        Set-StrictMode -Version 'Latest'
        Use-CallerPreference -Cmdlet $PSCmdlet -Session $ExecutionContext.SessionState

        function Test-ProgramInstalled
        {
            [CmdletBinding()]
            param(
                [Parameter(Mandatory)]
                [String] $Name,

                [Guid] $Code
            )

            $DebugPreference = 'SilentlyContinue'
            $installInfo = Get-InstalledProgram -Name $Name -ErrorAction Ignore
            if( -not $installInfo )
            {
                return $false
            }

            $installed = $installInfo.ProductCode -eq $Code
            if( $installed )
            {
                $msg = "$($msgPrefix)[$($installInfo.DisplayName)]  Installed $($installInfo.InstallDate)."
                Write-Verbose -Message $msg
                return $true
            }

            return $false
        }

        function Invoke-Msiexec
        {
            [CmdletBinding()]
            param(
                [Parameter(Mandatory, ValueFromPipeline)]
                [Object] $Msi,

                [String] $From
            )

            process
            {
                $target = $Msi.ProductName
                if( $Msi.Manufacturer )
                {
                    $target = "$($Msi.Manufacturer)'s ""$($target)"""
                }

                if( $Msi.ProductVersion )
                {
                    $target = "$($target) $($Msi.ProductVersion)"
                }

                $deleteLog = $false
                if( -not $LogPath )
                {
                    $LogPath = "$($Msi.Path | Split-Path -Leaf).$([IO.Path]::GetRandomFileName()).log"
                    $LogPath = Join-Path -Path ([IO.Path]::GetTempPath()) -ChildPath $LogPath
                    $deleteLog = $true
                }
                $logParentDir = $LogPath | Split-Path -Parent
                if( $logParentDir -and -not (Test-Path -Path $logParentDir) )
                {
                    New-Item -Path $logParentDir -Force -ItemType 'Directory' | Out-Null
                }
                if( -not (Test-Path -Path $LogPath) )
                {
                    New-Item -Path $LogPath -ItemType 'File' | Out-Null
                }

                if( -not $From )
                {
                    $From = $Msi.Path
                }

                $displayOptions = @{
                    'Quiet' = '/quiet';
                    'Passive' = '/passive';
                    'Full' = '';
                }

                $ArgumentList = & {
                    '/i'
                    # Must surround with double quotes. Single quotes are interpreted as part of the path.
                    """$($msi.Path)"""
                    $displayOptions[$DisplayMode]
                    $ArgumentList | Write-Output
                    "/l$($LogOption)",
                    # Must surround with double quotes. Single quotes are interpreted as part of the path.
                    """$($LogPath)"""
                } | Where-Object { $_ }

                $action = 'Install'
                $verb = 'Installing'
                if( $Force )
                {
                    $action = 'Repair'
                    $verb = 'Repairing'
                }

                if( $PSCmdlet.ShouldProcess( $From, $action ) )
                {
                    Write-Information -Message "$($msgPrefix)$($verb) $($target) from ""$($From)"""
                    Write-Verbose -Message "msiexec.exe $($ArgumentList -join ' ')"
                    $timer = [Diagnostics.Stopwatch]::StartNew()
                    $msiProcess = Start-Process -FilePath 'msiexec.exe' `
                                                -ArgumentList $ArgumentList `
                                                -NoNewWindow `
                                                -PassThru `
                                                -Wait

                    $timer.Stop()
                    $msg =
                        "            exit($($msiProcess.ExitCode)) in $($timer.Elapsed.TotalSeconds.ToString('0.000'))s"
                    Write-Verbose $msg

                    if( $null -ne $msiProcess.ExitCode -and $msiProcess.ExitCode -ne 0 )
                    {
                        $msg = "$($target) $($action.ToLowerInvariant()) failed. Installer ""$($msi.Path)"" returned " +
                               "exit code $($msiProcess.ExitCode). See the installation log file ""$($LogPath)"" for " +
                               'more information and https://docs.microsoft.com/en-us/windows/win32/msi/error-codes ' +
                               'for a description of the exit code.'
                        Write-Error $msg -ErrorAction $ErrorActionPreference
                        return
                    }
                }

                if( $deleteLog -and (Test-Path -Path $LogPath) )
                {
                    Remove-Item -Path $LogPath -ErrorAction Ignore
                }
            }
        }

        $msgPrefix = "[$($MyInvocation.MyCommand.Name)]  "
        Write-Debug "$($msgPrefix)+"
        
        if( $Path )
        {
            Get-Msi -Path $Path |
                Where-Object {
                    $msiInfo = $_

                    $installed = Test-ProgramInstalled -Name $msiInfo.ProductName -Code $msiInfo.ProductCode
                    if( $installed )
                    {
                        return $Force.IsPresent
                    }

                    # Not installed so $Force has no meaning. $Force also controls whether the action is "Install" or
                    # "Repair". We're always installing if not installed so set $Force to $false.
                    $Force = $false
                    return $true
                } |
                Invoke-Msiexec
            return
        }

        # If the program we are going to download is already installed, don't re-download it.
        $installed = Test-ProgramInstalled -Name $ProductName -Code $ProductCode
        if( $installed -and -not $Force )
        {
            return
        }
        # Make sure action is properly reported as Install or Repair.
        if( -not $installed )
        {
            $Force = $false
        }

        $msi = Get-Msi -Url $Url
        if( -not $msi )
        {
            return
        }

        $actualChecksum = Get-FileHash -LiteralPath $msi.Path
        if( $actualChecksum.Hash -ne $Checksum )
        {
            $msg = "Install failed: checksum ""$($actualChecksum.Hash.ToLowerInvariant())"" for installer " +
                   "downloaded from ""$($Url)"" does not match expected checksum ""$($Checksum.ToLowerInvariant())""."
            Write-Error -Message $msg -ErrorAction $ErrorActionPreference
            return
        }

        $msi | Invoke-Msiexec -From $Url

        Write-Debug "$($msgPrefix)-"
    }
}
